### PR TITLE
Various build/test/docs fixes and enhancements

### PR DESCRIPTION
**Pull request by dalcinl**

### CODE REVIEW

**Comment by dalcinl** [comment]
@leofang You may want to update issue #71  after this one is merged.

File: docs/source/install.rst
Comment (leofang) [inline docs/source/install.rst:51]: Should we deprecate it?

File: docs/source/install.rst
Comment (leofang) [inline docs/source/install.rst:55]:
```suggestion
   If the MPI implementation does not provide a compiler wrapper, or it
```

File: test/arrayimpl.py
Comment (leofang) [inline test/arrayimpl.py:530]: I am lost, what is `type(case)`?

File: test/arrayimpl.py
Comment (leofang) [inline test/arrayimpl.py:532]: and why would `loop.typecode` and `loop.array` be defined at this point?

File: test/arrayimpl.py
Comment (leofang) [inline test/arrayimpl.py:517]: Not a big fan of attaching attributes to a function... If `loop` has to have attributes, can't we just turn it into a class that has `__call__`?

File: docs/source/install.rst
Comment (leofang) [inline docs/source/install.rst:58]: nit:
```suggestion
   provided in an ini-style configuration file under a ``[mpi]``
```

File: docs/source/install.rst
Comment (leofang) [inline docs/source/install.rst:65]: This is handy 👍

File: docs/source/install.rst
Comment (dalcinl) [reply docs/source/install.rst:51]: I'm sick to the bones with Python's packaging-related deprecations and I don't want to contribute to this trend. Moreover, the main reason I'm adding the `MPI4PY_BUILD_MPICFG` name is to provide proper namespacing. I'm even considering whether we should support `MPI4PY_BUILD_MPICC` and so on, keeping the previous ones as aliases.

File: test/arrayimpl.py
Comment (dalcinl) [reply test/arrayimpl.py:532]: Because of the way `loop()` and `test()` are used. The point of these two routines, silly as they may look, it to reduce the verbosity of looping over arraytypes/typecodes and using the subTest context manager at the same time.

File: test/arrayimpl.py
Comment (dalcinl) [reply test/arrayimpl.py:517]: I use function attributes sparingly, as a way to track and hide things that would otherwise have to be stored in global variables. At this point, I do not want to complicate the implementation of all this silly extra testing support. If mpi4py would use a more advanced testing framework, then all this mess would probably not be there. But I do not want to depend on stuff outside Python's stdlib.

**Comment by dalcinl** [comment]
The testsuite is definitely not nice code. mpi4py testing is not a fun part of maintaining this project. I'm quite sick of it, actually. I'm totally tired of chasing issues related to poor development practices in projects mpi4py depends on but I'm not responsible of. Therefore, for the time being, I do my best to keep the stuff under `test/` working, but as it is not user-facing, I do not care too much about it being good looking.

File: docs/source/install.rst
Comment (leofang) [reply docs/source/install.rst:51]: Perhaps I wasn't being clear, I meant deprecate those env vars without a namespace scope. I am in favor of all env vars scoped by the project name to avoid potential issues.

**Review by leofang — APPROVED** [review]

File: docs/source/install.rst
Comment (dalcinl) [reply docs/source/install.rst:51]: I understood you. But what do you mean by "deprecation"? Emit deprecation warnings? That's what I'm not in favor of: more packaging/build deprecations, I'm sick of it. Without emitting a deprecation warning, the whole discussion is moot. But even if we emit a deprecation warning, I don't see any reason (besides naming aesthetics) for actually remove the short names with the obvious compatibility breakage.

In my view, there is nothing technically wrong with a short name like `MPICC`, it is more convenient for anyone installing manually in the commmand line, for example. The only reason I would like to add `MPI4PY_BUILD_MPICC` is to provide a project-specific name that can be used to avoid conflicts with other projects using the name `MPICC` with different semantics.

So, let the short names survive, and add namespaced names for all these vars. I'll push a new commit shortly.

File: docs/source/install.rst
Comment (dalcinl) [inline docs/source/install.rst:73]: @leofang Do you like it this way?

Comment (leofang) [reply docs/source/install.rst:73]: Thanks, Lisandro, LGTM.

**Review by leofang — APPROVED** [review]

File: docs/source/install.rst
Comment (leofang) [reply docs/source/install.rst]: Avoiding conflicts means everything to me and I don't really care aesthetics whatsoever. It's extremely painful to debug in days and finally realize some env vars could be accepted in multiple projects but they expect the same env var to be used in slightly different fashion. As a recent example, `NVCC` in some projects can be prefixed with `ccache` but some explicitly reject this idea. The outcome is the build pipeline failed with an obscure error. There's no way a user or package maintainer can figure it out without lengthy investigation.

**Pull request merged**
